### PR TITLE
JIT: don't clone loops where init or limit is a cast local

### LINKED ISSUE #57535: JIT crash/assert while compiling Grpc.core codebase
Repro instructions:
Download https://microsoft-my.sharepoint.com/:u:/p/davidwr/ETDI552L1RtCugwcniHGv5QBVsNi3PDCpXt6y-tWayhvow?e=f4wH4F and unzip it to a directory.

Using that directory as the working directory, run crossgen2 passing the rsp file in the zip as a response file. For example:

`c:\git\runtime\dotnet c:\git\runtime\artifacts\bin\coreclr\windows.x64.Debug\crossgen2\crossgen2.dll @crossgen2repro.rsp`

If run against a Debug crossgen2, it will produce an assert 
```
C:\git\runtime\src\coreclr\jit\loopcloning.cpp:183
Assertion failed 'genTypeSize(genActualType(op1Tree->TypeGet())) == genTypeSize(genActualType(op2Tree->TypeGet()))' in 'Grpc.Core.Internal.SliceBufferSafeHandle:ToByteArray():System.Byte[]:this' during 'Clone loops' (IL size 215)
```
If run against a release crossgen2 it will produce an AV
```
Fatal error. System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
Repeat 2 times:
--------------------------------
   at Internal.JitInterface.CorInfoImpl.JitCompileMethod(IntPtr ByRef, IntPtr, IntPtr, IntPtr, Internal.JitInterface.CORINFO_METHOD_INFO ByRef, UInt32, IntPtr ByRef, UInt32 ByRef)
--------------------------------
   at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(ILCompiler.DependencyAnalysis.IMethodNode, Internal.IL.MethodIL)
   at Internal.JitInterface.CorInfoImpl.CompileMethod(ILCompiler.DependencyAnalysis.ReadyToRun.MethodWithGCInfo, ILCompiler.Logger)
   at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__36_0(ILCompiler.DependencyAnalysisFramework.DependencyNodeCore`1<ILCompiler.DependencyAnalysis.NodeFactory>)
   at System.Threading.Tasks.Parallel+<>c__DisplayClass33_0`2[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e],[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForEachWorker>b__0(Int32)
   at System.Threading.Tasks.Parallel+<>c__DisplayClass19_0`1[[System.__Canon, System.Private.CoreLib, Version=6.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]].<ForWorker>b__1(System.Threading.Tasks.RangeWorker ByRef, Int32, Boolean ByRef)
   at System.Threading.Tasks.TaskReplicator+Replica.Execute()
   at System.Threading.Tasks.TaskReplicator+Replica+<>c.<.ctor>b__4_0(System.Object)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task+<>c.<.cctor>b__271_0(System.Object)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(System.Threading.Thread, System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef, System.Threading.Thread)
   at System.Threading.Tasks.Task.ExecuteEntryUnsafe(System.Threading.Thread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool+WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
```

### PR DESCRIPTION
**Pull request by AndyAyersMS**
The loop cloner assumes all computations it introduces are compatible
with TYP_INT, so don't allow cloning when the initial or final value
are variables with incompatible types.

Fixes #57535.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib @SingleAccretion 

No SPMI diffs.

We will want to port this to .NET 6.

File: src/coreclr/jit/loopcloning.cpp [hunk 1]
```diff
@@ -968,7 +968,14 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
         else if (loop->lpFlags & LPFLG_VAR_INIT)
         {
             // initVar >= 0
-            LC_Condition geZero(GT_GE, LC_Expr(LC_Ident(loop->lpVarInit, LC_Ident::Var)),
+            const unsigned initLcl = loop->lpVarInit;
+            if (!genActualTypeIsInt(lvaGetDesc(initLcl)->lvType))
```
Comment (SingleAccretion) [inline src/coreclr/jit/loopcloning.cpp:972]:
```suggestion
            if (!genActualTypeIsInt(lvaGetDesc(initLcl)))
```

File: src/coreclr/jit/loopcloning.cpp [hunk 2]
```diff
@@ -992,9 +999,14 @@ bool Compiler::optDeriveLoopCloningConditions(unsigned loopNum, LoopCloneContext
         }
         else if (loop->lpFlags & LPFLG_VAR_LIMIT)
         {
-            unsigned limitLcl = loop->lpVarLimit();
-            ident             = LC_Ident(limitLcl, LC_Ident::Var);
+            const unsigned limitLcl = loop->lpVarLimit();
+            if (!genActualTypeIsInt(lvaGetDesc(limitLcl)->lvType))
```
Comment (SingleAccretion) [inline src/coreclr/jit/loopcloning.cpp:1003]:
```suggestion
            if (!genActualTypeIsInt(lvaGetDesc(limitLcl)))
```

**Review by kunalspathak — APPROVED** [review]
LGTM. I am surprised that none of our test had this scenario.

**Review by SingleAccretion — COMMENTED** [review]
Question: do we think this is a cloning-exclusive issue? In relation to not setting the relevant loop flags in the first place (somewhere around `optRecordLoop`).

I suppose literally nothing but cloning uses this information today so it does not matter either way.

**Comment by AndyAyersMS** [comment]
> Question: do we think this is a cloning-exclusive issue? In relation to not setting the relevant loop flags in the first place (somewhere around `optRecordLoop`).
> 
> I suppose literally nothing but cloning uses this information today so it does not matter either way.

Probably best to keep the fix surgical for .NET 6.

But we should look and see if anything else might be tripped up.

**Comment by AndyAyersMS** [comment]
Linux failures seems like some sort of test issue with crossgen2 testing:
```
;; crossgen2determinism.sh

Unhandled exception. System.AggregateException: One or more errors occurred. (Code generation failed for method '[crossgen2smoke]Program.TestSES_NestedU8(SingleElementStruct_NestedU8,object)') (Code generation failed for method '[crossgen2smoke]Program.TestSES_NestedR8(SingleElementStruct_NestedR8,object)')
---> ILCompiler.CodeGenerationFailedException: Code generation failed for method '[crossgen2smoke]Program.TestSES_NestedU8(SingleElementStruct_NestedU8,object)'
---> System.MissingMethodException: Method not found: 'Void System.Diagnostics.Debug.Assert(Boolean, AssertInterpolatedStringHandler)'.
```
@trylek seen anything like this recently?

**Comment by hoyosjs** [comment]
@AndyAyersMS yes, it's been reported in a few cases. Issue: https://github.com/dotnet/runtime/issues/57620

**Comment by AndyAyersMS** [comment]
Mono interpreter failure looks real but is unrelated.
```
   Loader.classloader.XUnitWrapper: [Long Running Test] 'Loader/classloader/regressions/523654/test532654_b/test532654_b.sh', Elapsed: 00:08:08
Attempting to collect crash dump: /home/helixbot/dotnetbuild/dumps/crashdump_269.dmp
Invoking: sudo /root/helix/work/correlation/createdump --name "/home/helixbot/dotnetbuild/dumps/crashdump_269.dmp" 269 --withheap
createdump stdout:
Gathering state for process 269 corerun

createdump stderr:
CLRDataCreateInstance(ICLRDataEnumMemoryRegions) FAILED 80131c4f

Failed to collect crash dump
    Loader/classloader/regressions/523654/test532654_b/test532654_b.sh [FAIL]
      corrupted double-linked list

        Got a SIGABRT while executing native code. This usually indicates
        a fatal error in the mono runtime or one of the native libraries 
        used by your application
```
Does not seem to be a known issue? cc @BrzVlad

**Pull request merged**